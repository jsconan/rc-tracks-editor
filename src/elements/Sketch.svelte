<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import defined from '../helpers/defined.js';

    export let x = 0;
    export let y = 0;
    export let viewWidth = void 0;
    export let viewHeight = void 0;
    export let width = void 0;
    export let height = void 0;

    const hasViewBox = () =>
        (defined(viewWidth) || Number.isFinite(width)) && (defined(viewHeight) || Number.isFinite(height));

    $: viewBox = hasViewBox() ? `${x} ${y} ${viewWidth || width} ${viewHeight || height}` : void 0;
</script>

<svg {viewBox} {width} {height}>
    <defs>
        <slot name="defs" />
    </defs>
    <slot />
</svg>
