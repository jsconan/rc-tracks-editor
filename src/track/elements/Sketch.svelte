<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { defined } from '../../core/helpers';

    let cls = void 0;
    export { cls as class };
    export let style = void 0;
    export let id = void 0;
    export let x = void 0;
    export let y = void 0;
    export let width = void 0;
    export let height = void 0;
    export let viewX = 0;
    export let viewY = 0;
    export let viewWidth = void 0;
    export let viewHeight = void 0;

    const hasViewBox = () =>
        (defined(viewWidth) || Number.isFinite(width)) && (defined(viewHeight) || Number.isFinite(height));

    $: viewBox = hasViewBox() ? `${viewX} ${viewY} ${viewWidth || width} ${viewHeight || height}` : void 0;
</script>

<svg {id} class={cls} {style} {x} {y} {viewBox} {width} {height} on:click>
    {#if $$slots.defs}
        <defs>
            <slot name="defs" />
        </defs>
    {/if}
    <slot />
</svg>
