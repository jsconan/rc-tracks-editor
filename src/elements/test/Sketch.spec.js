/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from '@testing-library/svelte';
import Sketch from '../Sketch.svelte';
import SketchWithSlot from './SketchWithSlot.svelte';

/**
 * Resolves a promise after a particular number of milliseconds.
 * @param {number} ms - How many milliseconds to wait.
 * @returns {Promise}
 */
const wait = ms => new Promise(resolve => setTimeout(resolve, ms));

const x = 100;
const y = 150;
const viewWidth = 200;
const viewHeight = 300;
const width = 100;
const height = 150;

describe('Sketch', () => {
    it('renders with default values', () => {
        const { container } = render(Sketch);

        expect(container).toMatchSnapshot();
    });

    it.each([
        ['width and height', { width, height }],
        ['x, y, width, and height', { x, y, width, height }],
        ['x, y, viewWidth, and viewHeight', { x, y, viewWidth, viewHeight }],
        ['x, y, viewWidth, viewHeight, width, and height', { x, y, viewWidth, viewHeight, width, height }]
    ])('renders with the parameters %s', (title, props) => {
        const { container } = render(Sketch, { props });

        expect(container).toMatchSnapshot();
    });

    it.each([
        ['width and height', { width, height }],
        ['x, y, width, and height', { x, y, width, height }],
        ['x, y, viewWidth, and viewHeight', { x, y, viewWidth, viewHeight }],
        ['x, y, viewWidth, viewHeight, width, and height', { x, y, viewWidth, viewHeight, width, height }]
    ])('updates with the parameters %s', (title, props) => {
        const unset = {
            x: void 0,
            y: void 0,
            viewWidth: void 0,
            viewHeight: void 0,
            width: void 0,
            height: void 0
        };
        const rendered = render(Sketch);

        return wait(50)
            .then(() => rendered.component.$set(props))
            .then(() => expect(rendered.container).toMatchSnapshot())
            .then(() => wait(50))
            .then(() => rendered.component.$set(unset))
            .then(() => expect(rendered.container).toMatchSnapshot());
    });

    it('renders with the given element in slots', () => {
        const { container } = render(SketchWithSlot);

        expect(container).toMatchSnapshot();
    });
});
