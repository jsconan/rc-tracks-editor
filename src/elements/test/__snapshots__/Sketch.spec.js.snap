// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StraightTile renders with default values 1`] = `
<body>
  <div>
    <svg>
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile renders with the given element in slots 1`] = `
<body>
  <div>
    <svg>
      <defs>
        <circle
          cx="5"
          cy="5"
          id="circle"
          r="10"
          slot="defs"
        />
      </defs>
      <rect
        height="50"
        width="100"
        x="10"
        y="10"
      />
    </svg>
  </div>
</body>
`;

exports[`StraightTile renders with the parameters width and height 1`] = `
<body>
  <div>
    <svg
      height="150"
      viewBox="0 0 100 150"
      width="100"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile renders with the parameters x, y, viewWidth, and viewHeight 1`] = `
<body>
  <div>
    <svg
      viewBox="100 150 200 300"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile renders with the parameters x, y, viewWidth, viewHeight, width, and height 1`] = `
<body>
  <div>
    <svg
      height="150"
      viewBox="100 150 200 300"
      width="100"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile renders with the parameters x, y, width, and height 1`] = `
<body>
  <div>
    <svg
      height="150"
      viewBox="100 150 100 150"
      width="100"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters width and height 1`] = `
<body>
  <div>
    <svg
      height="150"
      viewBox="0 0 100 150"
      width="100"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters width and height 2`] = `
<body>
  <div>
    <svg>
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters x, y, viewWidth, and viewHeight 1`] = `
<body>
  <div>
    <svg
      viewBox="100 150 200 300"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters x, y, viewWidth, and viewHeight 2`] = `
<body>
  <div>
    <svg>
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters x, y, viewWidth, viewHeight, width, and height 1`] = `
<body>
  <div>
    <svg
      height="150"
      viewBox="100 150 200 300"
      width="100"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters x, y, viewWidth, viewHeight, width, and height 2`] = `
<body>
  <div>
    <svg>
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters x, y, width, and height 1`] = `
<body>
  <div>
    <svg
      height="150"
      viewBox="100 150 100 150"
      width="100"
    >
      <defs />
    </svg>
  </div>
</body>
`;

exports[`StraightTile updates with the parameters x, y, width, and height 2`] = `
<body>
  <div>
    <svg>
      <defs />
    </svg>
  </div>
</body>
`;
